File: export_data.js
import fs from 'fs';
import { execSync } from 'child_process';

const schemas = ['auth', 'storage'];
const tables = {
  auth: ['users', 'identities'],
  storage: ['buckets', 'objects']
};

function formatValue(val) {
  if (val === null) return 'NULL';
  if (typeof val === 'string') return `'${val.replace(/'/g, "''")}'`;
  if (typeof val === 'object') return `'${JSON.stringify(val).replace(/'/g, "''")}'`;
  return val;
}

async function exportTable(schema, table) {
  console.log(`Exporting ${schema}.${table}...`);
  const query = `SELECT * FROM ${schema}.${table}`;
  const result = execSync(`psql -t -c "COPY (${query}) TO STDOUT WITH (FORMAT CSV, HEADER)"`).toString();
  
  // Actually, using COPY is easier if I want to import it later with COPY, 
  // but the user asked for SQL INSERTs or migration files.
  // Let's use psql to generate INSERT statements.
}

File: export_data.py
import json
import subprocess
import sys

def fetch_table(schema, table):
    # Fetching row by row to avoid memory issues and handle JSON correctly
    cmd = ["psql", "-t", "-c", f"SELECT row_to_json(t) FROM (SELECT * FROM {schema}.{table}) t;"]
    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        print(f"Error fetching {schema}.{table}: {result.stderr}", file=sys.stderr)
        return []
    lines = result.stdout.strip().split('\n')
    data = []
    for line in lines:
        if line.strip():
            data.append(json.loads(line))
    return data

def generate_inserts(schema, table, data):
    if not data: return f"-- No data for {schema}.{table}\n"
    columns = list(data[0].keys())
    sql = f"-- Data for {schema}.{table}\n"
    for row in data:
        vals = []
        for col in columns:
            val = row[col]
            if val is None:
                vals.append("NULL")
            elif isinstance(val, (dict, list)):
                json_str = json.dumps(val).replace("'", "''")
                vals.append(f"'{json_str}'")
            else:
                str_val = str(val).replace("'", "''")
                vals.append(f"'{str_val}'")
        sql += f"INSERT INTO {schema}.{table} ({', '.join(columns)}) VALUES ({', '.join(vals)});\n"
    return sql

def export_schema_data(schema, tables, filename):
    with open(f"/mnt/documents/{filename}", "w") as f:
        f.write("BEGIN;\n")
        f.write("SET session_replication_role = 'replica';\n")
        for table in tables:
            data = fetch_table(schema, table)
            f.write(generate_inserts(schema, table, data))
        f.write("SET session_replication_role = 'origin';\n")
        f.write("COMMIT;\n")

# Export Auth
export_schema_data("auth", ["users", "identities"], "auth_data.sql")
# Export Storage Metadata
export_schema_data("storage", ["buckets", "objects"], "storage_metadata.sql")

